nemo/collections/nlp/models/entity_linking/entity_linking_model.py

```diff
@@ -0,0 +1,193 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, Optional
+
+import torch
+from omegaconf import DictConfig
+from pytorch_lightning import Trainer
+from transformers import AutoTokenizer
+
+from nemo.collections.common.losses import MultiSimilarityLoss
+from nemo.collections.nlp.data import EntityLinkingDataset
+from nemo.collections.nlp.models.nlp_model import NLPModel
+from nemo.collections.nlp.modules.common.lm_utils import get_lm_model
+from nemo.core.classes.common import typecheck
+from nemo.core.classes.exportable import Exportable
+from nemo.core.neural_types import ChannelType, LogitsType, MaskType, NeuralType
+from nemo.utils import logging
+
+__all__ = ['EntityLinkingModel']
+
+
+class EntityLinkingModel(NLPModel, Exportable):
+    """
+    Second stage pretraining of BERT based language model
+    for entity linking task.
+    """
+
+    @property
+    def input_types(self) -> Optional[Dict[str, NeuralType]]:
+        return {
+            "input_ids": NeuralType(('B', 'T'), ChannelType()),
+            "attention_mask": NeuralType(('B', 'T'), MaskType(), optional=True),
+            "token_type_ids": NeuralType(('B', 'T'), ChannelType(), optional=True),
+        }
+
+    @property
+    def output_types(self) -> Optional[Dict[str, NeuralType]]:
+        return {"logits": NeuralType(('B', 'D'), LogitsType())}
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer = None):
+        """Initializes the SAP-BERT model for entity linking."""
+
+        # tokenizer needed before super().__init__() so dataset and loader can process data
+        self._setup_tokenizer(cfg.tokenizer)
+
+        super().__init__(cfg=cfg, trainer=trainer)
+
+        self.model = get_lm_model(
+            pretrained_model_name=cfg.language_model.pretrained_model_name,
+            config_file=cfg.language_model.config_file,
+            config_dict=cfg.language_model.config,
+            checkpoint_file=cfg.language_model.lm_checkpoint,
+        )
+
+        # Token to use for the self-alignment loss, typically the first token, [CLS]
+        self._idx_conditioned_on = 0
+        self.loss = MultiSimilarityLoss()
+        self.index = None
```

Comment:
Lets see if we can try and register the index as an artifact so it gets saved in the `.nemo` file

Reply:
I actually think index doesn't need to be registered as an artifact. Its optionally created after the model is fully trained and saved. I removed the index attribute from the entity linking encoder.